ProjectCoimbra.UWP/Project.Coimbra.Midi/MidiEngine.cs

```diff
@@ -241,18 +241,37 @@ private static string GetTrackDisplayName(StorageFile file, MidiFile midi)
 
         private void AddInstruments(MidiFile midi)
         {
-            foreach (var programChangeEvent in midi.GetTimedEvents().Select(e => e.Event).OfType<ProgramChangeEvent>())
+            var timedEvents = midi.GetTimedEvents();
+            var programChangedEvents = timedEvents.Select(e => e.Event).OfType<ProgramChangeEvent>().ToList<ChannelEvent>();
```

Comment:
It looks like you don't use `programChangedEvents` in many places. I believe you could skip this entirely and just apply the result of this operation to `events`.